### ISSUE TITLE
Running final binary as sudo

### ISSUE THREAD

**Issue by alexzk1** [doe300/VC4CL#41]
...and getting

 

> INFO:0] Initialize OpenCL runtime...
> [ INFO:0] Successfully initialized OpenCL cache directory: /root/.cache/opencv/3.4.1/opencl_cache/
> [ INFO:0] Preparing OpenCL cache configuration for context: 32-bit--Broadcom--VideoCore_IV_GPU--0_4
> OpenCV(3.4.1) Error: Unknown error code -220 (OpenCL error CL_INVALID_VALUE (-30) during call: clGetProgramInfo(handle, CL_PROGRAM_BINARIES, sizeof(ptr), &ptr, NULL)) in getProgramBinary, file /build/opencv/src/opencv-3.4.1/modules/core/src/ocl.cpp, line 3752
> [ WARN:0] Can't save OpenCL binary into cache: /root/.cache/opencv/3.4.1/opencl_cache/32-bit--Broadcom--VideoCore_IV_GPU--0_4/imgproc--filterSepRow_e99b92fca8604fe253f3c641802ce117.bin
> OpenCV(3.4.1) /build/opencv/src/opencv-3.4.1/modules/core/src/ocl.cpp:3752: error: (-220) OpenCL error CL_INVALID_VALUE (-30) during call: clGetProgramInfo(handle, CL_PROGRAM_BINARIES, sizeof(ptr), &ptr, NULL) in function getProgramBinary
> 
> OpenCL error CL_INVALID_WORK_GROUP_SIZE (-54) during call: clEnqueueNDRangeKernel('row_filter_C1_D0', dims=2, globalsize=240x544x1, localsize=16x16x1) sync=false
> OpenCV(3.4.1) Error: Unknown error code -220 (OpenCL error CL_INVALID_VALUE (-30) during call: clGetProgramInfo(handle, CL_PROGRAM_BINARIES, sizeof(ptr), &ptr, NULL)) in getProgramBinary, file /build/opencv/src/opencv-3.4.1/modules/core/src/ocl.cpp, line 3752
> [ WARN:0] Can't save OpenCL binary into cache: /root/.cache/opencv/3.4.1/opencl_cache/32-bit--Broadcom--VideoCore_IV_GPU--0_4/imgproc--filterSepRow_e99b92fca8604fe253f3c641802ce117.bin
> OpenCV(3.4.1) /build/opencv/src/opencv-3.4.1/modules/core/src/ocl.cpp:3752: error: (-220) OpenCL error CL_INVALID_VALUE (-30) during call: clGetProgramInfo(handle, CL_PROGRAM_BINARIES, sizeof(ptr), &ptr, NULL) in function getProgramBinary
> 
> OpenCL error CL_INVALID_WORK_GROUP_SIZE (-54) during call: clEnqueueNDRangeKernel('row_filter_C1_D0', dims=2, globalsize=240x544x1, localsize=16x16x1) sync=false
> FPS  0.0205593, Objects: 0
> OpenCL error CL_INVALID_WORK_GROUP_SIZE (-54) during call: clEnqueueNDRangeKernel('row_filter_C1_D0', dims=2, globalsize=240x544x1, localsize=16x16x1) sync=false
> OpenCL error CL_INVALID_WORK_GROUP_SIZE (-54) during call: clEnqueueNDRangeKernel('row_filter_C1_D0', dims=2, globalsize=240x544x1, localsize=16x16x1) sync=false

Is it problem of what ? >: Not supported features, rights, opencl?
Made just in case
>sudo chmod 777 /root/.cache/opencv/3.4.1/opencl_cache/32-bit--Broadcom--VideoCore_IV_GPU--0_4

 and didnt work

**Comment by doe300** [comment]
You get two different error codes:

> OpenCL error CL_INVALID_WORK_GROUP_SIZE (-54)

OpenCV seems to not be able to handle small work-group sizes, or at least has a lower bound larger than the supported work-group size of VC4CL.

> Error: Unknown error code -220

This seem to be that the call to `clGetProgramInfo` failed with somehow both the error code `CL_INVALID_VALUE (-30)` and an unknown error code `-220`.

**Comment by alexzk1** [comment]
Fails here:Mailbox::executeQPU on sending ioctl, everything hangs
as 

> [VC4CL] Running work-group 0, 0, 0
> [VC4CL] Mailbox buffer before:
> [VC4CL] 0000: 0x00000028
> [VC4CL] 0004: 0x00000000
> [VC4CL] 0008: 0x00030011
> [VC4CL] 000c: 0x00000010
> [VC4CL] 0010: 0x00000010
> [VC4CL] 0014: 0x0000000c
> [VC4CL] 0018: 0x5e10df58
> [VC4CL] 001c: 0x00000000
> [VC4CL] 0020: 0x00007530
> [VC4CL] 0024: 0x751687cc
> 
Code is:
 R"CLC(
        __kernel void TestMagic(const int total, const int is_deeper_magic, const float alpha_s, const float fore_th, __global const float* gradx, __global const float* grady,
                                                 //in/out
                                                 __global float* BSx,  __global float* BSy, __global int* mapRes)
        {
            private const size_t i        = get_global_id(0);
            private const size_t gpu_used = get_global_size(0);

            private const size_t elements_count = total / (gpu_used * 16);
            private const size_t offset = i * total / gpu_used;

            for (size_t k = 0; k < elements_count; ++k)
            {
               int16 mr           = vload16( k , mapRes + offset);
               const int16 twos   = 2;
               mr += twos;
               vstore16(mr, k, mapRes + offset);
            }
        }
        )CLC",
> 
During compilation:

> [VC4CL] Precompiling source with: 
> [VC4CL] Precompilation complete with status: 0
> [VC4CL] Compilation log: [W] Sat Jul  7 03:04:50 2018: Warnings in precompilation:
> [W] Sat Jul  7 03:04:50 2018: <stdin>:90:9: warning: null character ignored
>         <U+0000>
>         ^
> 1 warning generated.

Any ideas? >: QPULib examples work...(https://github.com/mn416/QPULib)
>

**Comment by alexzk1** [comment]
![img_2906](https://user-images.githubusercontent.com/4589845/42405953-392daf5e-81a7-11e8-9c9d-1b5677eb3c11.JPG)

**Comment by doe300** [comment]
Sometimes, if the code generated for a kernel does something completely wrong, then the QPUs get into a hanged state. Looks like this is happening here

**Comment by alexzk1** [comment]
and how to deal with? is it compiler problem or my code?

**Comment by doe300** [comment]
Most likely a compiler problem. I found an issue which also occurs in the kernel given, and will try to fix it.

**Comment by alexzk1** [comment]
what exactly ? I could change my kernel maybe to skip this problem.

**Comment by alexzk1** [comment]
This commit does not fix issue, still hangs
https://github.com/doe300/VC4C/commit/f13120debba4d59da6b8ca432a22eaf64e338550

**Comment by doe300** [comment]
Yeah, the integer division (and modulo) is still wrong. I could not figure out the reason for this yesterday.

**Comment by alexzk1** [comment]
does it mean float div will work? I need that for atan2, which contains same integer div (for offset) and inside atan2 y/x ... btw in you souce it is some comment there, maybe you do x/y ?

**Comment by doe300** [comment]
Floating point division works, at least for the tests I ran, with worse-than-allowed accuracy in some cases. I don't know if the atan2 function works..

**Comment by alexzk1** [comment]
Yes, thats weird. Opencl on nvidia give a bit different results, then opencv-default(cpu?) does:
1st is opencv-default, 2nd is opencl-nvidia kernel:

> DUMP: 
> x = 0; y = 2; a = 1.5708
> x = -4; y = 4; a = 2.35636
> x = 6; y = 8; a = 0.927403
> x = 21; y = 13; a = 0.554224
> x = 18; y = 16; a = 0.726774
> x = 12; y = 6; a = 0.463683
> x = 15; y = -13; a = 5.569
> x = 8; y = -22; a = 5.06123
> x = -9; y = -11; a = 4.02667
> x = -11; y = 11; a = 2.35636
> x = -2; y = 18; a = 1.68144
> x = 4; y = 4; a = 0.785232
> x = 4; y = -12; a = 5.0342
> x = 7; y = -13; a = 5.20633
> x = 8; y = 0; a = 0
> x = 0; y = 10; a = 1.5708
> DUMP: 
> x = 0; y = 2; a = 1.5708
> x = -4; y = 4; a = 2.35619
> x = 6; y = 8; a = 0.927295
> x = 21; y = 13; a = 0.554308
> x = 18; y = 16; a = 0.726642
> x = 12; y = 6; a = 0.463648
> x = 15; y = -13; a = 5.56909
> x = 8; y = -22; a = 5.06116
> x = -9; y = -11; a = 4.02666
> x = -11; y = 11; a = 2.35619
> x = -2; y = 18; a = 1.68145
> x = 4; y = 4; a = 0.785398
> x = 4; y = -12; a = 5.03414
> x = 7; y = -13; a = 5.20633
> x = 8; y = 0; a = 0
> x = 0; y = 10; a = 1.5708

**Comment by alexzk1** [comment]
Used nvidia reference implementation (http://developer.download.nvidia.com/cg/atan2.html),on desktop it's ok. 

```
float16 myatan2(float16 y, float16 x)
        {
          float16 t0, t1, t2, t3, t4;

          t3 = fabs(x);
          t1 = fabs(y);
          t0 = max(t3, t1);
          t1 = min(t3, t1);
          t3 = 1.f / t0;
          t3 = t1 * t3;

          t4 = t3 * t3;
          t0 =         - 0.013480470f;
          t0 = t0 * t4 + 0.057477314f;
          t0 = t0 * t4 - 0.121239071f;
          t0 = t0 * t4 + 0.195635925f;
          t0 = t0 * t4 - 0.332994597f;
          t0 = t0 * t4 + 0.999995630f;
          t3 = t0 * t3;

          t3 = (fabs(y) > fabs(x)) ? 1.570796327f - t3 : t3;
          t3 = (x < 0) ?  3.141592654f - t3 : t3;
          t3 = (y < 0) ? -t3 : t3;

          return t3;
        }
```
On RPI it works as well:

> DUMP: 
> x = 0; y = 2; a = 1.5708
> x = -4; y = 4; a = 2.35636
> x = 6; y = 8; a = 0.927403
> x = 21; y = 13; a = 0.554224
> x = 18; y = 16; a = 0.726774
> x = 12; y = 6; a = 0.463683
> x = 15; y = -13; a = 5.569
> x = 8; y = -22; a = 5.06123
> x = -9; y = -11; a = 4.02667
> x = -11; y = 11; a = 2.35636
> x = -2; y = 18; a = 1.68144
> x = 4; y = 4; a = 0.785232
> x = 4; y = -12; a = 5.0342
> x = 7; y = -13; a = 5.20633
> x = 8; y = 0; a = 0
> x = 0; y = 10; a = 1.5708
> DUMP: 
> x = 0; y = 2; a = 1.5708
> x = -4; y = 4; a = 2.3562
> x = 6; y = 8; a = 0.927294
> x = 21; y = 13; a = 0.554308
> x = 18; y = 16; a = 0.72664
> x = 12; y = 6; a = 0.463646
> x = 15; y = -13; a = -0.714088
> x = 8; y = -22; a = -1.22203
> x = -9; y = -11; a = -2.25652
> x = -11; y = 11; a = 2.3562
> x = -2; y = 18; a = 1.68145
> x = 4; y = 4; a = 0.785395
> x = 4; y = -12; a = -1.24905
> x = 7; y = -13; a = -1.07686
> x = 8; y = 0; a = 0
> x = 0; y = 10; a = 1.5708
>

**Comment by doe300** [comment]
> Opencl on nvidia give a bit different results, then opencv-default

Depending on the compilation flags you specified, the NVIDIA code may use faster but inaccurate operations (e.g. due to `-cl-fast-relaxed-math` or `-cl-mad-enable`).

> On RPI something is broken completely:

This looks like the `atan2` function yielding wrong results. Some/a lot of the math functions in VC4CL are not correct or properly tested.

> Used nvidia reference implementation [...]

Thanks for the link, this might come in very handy, if I can figure out its license...

**Comment by alexzk1** [comment]
Still can't make "select" work >: is it broken too ? Bcs I was dependant on in another kernel as well (that initial I simplified for test).

**Comment by alexzk1** [comment]
Ok made custom select for float16, for integers must be even simplier:

```
 float16 myselect(float16 afalse, float16 atrue, int16 condition)
        {
            //we have -1 = true in condition ...it should be so
            float16 cond = convert_float16(condition) * -1.f;
            float16 not_cond = 1.f - cond;
            return atrue * cond + afalse * not_cond;
        }
```

**Comment by doe300** [comment]
You need to re-trigger building of the PCH and BC for vc4cl-stdlib, which is done by the script located in `${CMAKE_BINARY_DIR}/build/postinst` from the VC4C project for installed stdlib and `make vc4cl-stdlib` for stdlib sources (in which case you have to delete `<VC4CLStdLib-root>/include/VC4CLStdLib.h.pch` and `<VC4CLStdLib-root>/include/VC4CLStdLib.bc`).

**Comment by alexzk1** [comment]
Ok..so no compiler rebuild if pch is built as separated package like here https://github.com/alexzk1/vc4_stdlib_arch/blob/master/PKGBUILD ?

Btw, building VC4C gives many such warnings, is it ok ? (gcc 8)
usr/include/c++/8.1.0/bits/stl_vector.h: In member function 'std::vector<exprValue> Parser::parseArgumentList(const string&, size_t)':
/usr/include/c++/8.1.0/bits/stl_vector.h:1085:4: note: parameter passing for argument of type '__gnu_cxx::__normal_iterator<exprValue*, std::vector<exprValue> >' changed in GCC 7.1
In file included from /usr/include/c++/8.1.0/vector:69,

**Comment by doe300** [comment]
Well, the warnings are both libstdc++ internal warnings. I can try to disable the warning, but I don't think I can do anything about fixing them.

**Comment by alexzk1** [comment]
fmod(float16, float16) is still broken - pi hangs.

**Comment by alexzk1** [comment]
![img_2908](https://user-images.githubusercontent.com/4589845/42704499-af6f2f36-86d8-11e8-94c6-0cc04ce335e4.JPG)
No kernel panic, but that red dots around... maybe because connected monitor? And hang

**Comment by alexzk1** [comment]
holdon...why it says "myfabs", I have removed that func and it's not on disk at all in sources.

**Comment by alexzk1** [comment]
Did such macro, hanged  immediately without any 1 success step

>  #define Z1 ((float16)(a, b.s0123, b.s456789ab, b.scde))
>         #define Z2 (b)
>         #define Z3 ((float16) (b.s123, b.s4567, b.s89abcdef, c))
>         #define Z4 ((float16)(d, e.s0123, e.s456789ab, e.scde))
>         #define Z5 (e)
>         #define Z6 ((float16)(e.s123, e.s4567, e.s89abcdef, f))
>         #define Z7 ((float16)(g, h.s0123, h.s456789ab, h.scde))
>         #define Z8 (h)
>         #define Z9 ((float16)(h.s123, h.s4567, h.s89abcdef, i))

**Comment by alexzk1** [comment]
Okey, this seems working - function calls without macroses as parameters, code is wrong, but works, so how is it possible? >:

```
 p1 = myselectf16(p1, p2, atest);
        p2 = myselectf16(p2, p1, atest);

        atest =  isless(fabs(angle - pi4), pi8); //45
        p1 = myselectf16(p1, p2, atest);
        p2 = myselectf16(p2, p1, atest);

        atest =  islessequal(angle, pi8) || islessequal(fabs(angle - pi1), pi8); //0
        p1 = myselectf16(p1, p2, atest);
        p2 = myselectf16(p2, p1, atest);


        atest =  isless(fabs(angle - pi34), pi8); //135
        p1 = myselectf16(p1, p2, atest);
        p2 = myselectf16(p2, p1, atest);

```


In another kernel I had usage of those Z's in ariphmetics like
float16 Gx = (Z7 + 2 * Z8 + Z9) - (Z1 + 2 * Z2 +Z3);

and it is working ok.

**Comment by doe300** [comment]
> And maybe something else, replaced fmod by equivalent, but...on next run it was ok, and next-next run hanged again...

Did you reboot your Pi in between runs? Sometimes when the GPU hangs, it can only recovered by rebooting it.

> No kernel panic, but that red dots around... maybe because connected monitor?

It is possible, that the graphics driver collides with VC4CL, at least if VC4CL is compiled with `REGISTER_POKE_KERNELS` on. If the flag is disabled, the VC4CL uses the kernel synchronization routines, the graphics driver uses too (but see #15, idk if this is still valid).

> Okey, this seems working - function calls without macroses as parameters, code is wrong, but works, so how is it possible?

One theory: There is something wrong with vector shuffling (putting together vector from pieces of other vector), which you do in the macros.

**Comment by alexzk1** [comment]
> One theory: There is something wrong with vector shuffling (putting together vector from pieces of other vector), which you do in the macros.

Yes, possibly. Assignign like float16  z1 = Z1; and using variable didnt work too. However as I said other kernels were working. But not sure if gave correct results actually.

**Comment by alexzk1** [comment]
Last commits (from 15th Jul) to compiler/std lib do not fix hangs. However.. I see (always saw actually) 4 fps reports which is done once per 10 frames, so kernel still processes at least 40 frames. I use static buffers in C++code to avoid copies/creations. Is it possibly some error happens in opencl memory managment? >:

**Comment by doe300** [comment]
So what does the "40 frames" it runs mean? How are they split up in work-groups and work-items?
And do you run every 10 frames separately (e.g. with a new call to `clEnqueueNDKernel` or all at once?

Also, what do you mean by "[...]static buffers in C++code to avoid copies/creations"? You at least once have to copy the data to the GPU memory.

**Comment by alexzk1** [comment]
Here, that is last run i did:
https://github.com/alexzk1/LitterBug-Algorithm/blob/Desktop_Litterbug/cl/kernels.cpp

It runs about 40 times openCl::sobel2magic
which currently start non_maximum kernel only for testing. For example, sobel kernel runs fine many times (if uncomented)

**Comment by alexzk1** [comment]
So, I replaced all those Z's by e (it is Z5) and compiled kernel worked...
Then I replaced all by Z5, and it works again.
Then I tried those and still hangs:

```
 #define Z1 ((float16)(a, b.s0, b.s1, b.s2, b.s3, b.s4, b.s5, b.s6, b.s7, b.s8, b.s9, b.sa, b.sb, b.sc, b.sd, b.se))
        #define Z2 (b)
        #define Z3 ((float16) (b.s1, b.s2, b.s3, b.s4, b.s5, b.s6, b.s7, b.s8, b.s9, b.sa, b.sb, b.sc, b.sd, b.se, b.sf, c))
        #define Z4 ((float16)(d, e.s0, e.s1, e.s2, e.s3, e.s4, e.s5, e.s6, e.s7, e.s8, e.s9, e.sa, e.sb, e.sc, e.sd, e.se))
        #define Z5 (e)
        #define Z6 ((float16)(e.s1, e.s2, e.s3, e.s4, e.s5, e.s6, e.s7, e.s8, e.s9, e.sa, e.sb, e.sc, e.sd, e.se, e.sf, f))
        #define Z7 ((float16)(g, h.s0, h.s1, h.s2, h.s3, h.s4, h.s5, h.s6, h.s7, h.s8, h.s9, h.sa, h.sb, h.sc, h.sd, h.se))
        #define Z8 (h)
        #define Z9 ((float16)(h.s1, h.s2, h.s3, h.s4, h.s5, h.s6, h.s7, h.s8, h.s9, h.sa, h.sb, h.sc, h.sd, h.se, h.sf, i))

```

**Comment by doe300** [comment]
I am currently working on writing tests for and fixing most of the standard-library functions. But I will have a look at the vector insertion/extraction today and see if I can figure out, why the macros do not work.

**Comment by alexzk1** [comment]
Not sure if it is macros-only problem, another kernel works with:

> float16 Gx = (Z7 + 2 * Z8 + Z9) - (Z1 + 2 * Z2 + Z3);
> float16 Gy = (Z3 + 2 * Z6 + Z9) - (Z1 + 2 * Z4 + Z7);
>

**Comment by doe300** [comment]
I also tested the vector-assembly similar to what the macros do and they work correctly, so this is not the issue. Also, it would be very strange if the question whether it hangs depends on the values (whether correct or not).

One thing you could try is to compile the kernels with different optimization levels (`-O0`, `-O1` or`-O2`). This could give us a hint whether the error is in one of the optimizations or a general problem. Btw, the default level is `-O2` so all stable/halfway fast optimizations are activated.

**Comment by alexzk1** [comment]
last time i used was
progs.build("-cl-opt-disable");
can u check my sources link (and see history of the tries) ?
bcs i'm lost. I'm not expert in paraller or whatever, but ...i try my best, and check 1st of all on desktop/nvidia - when it works (no warnings) and does proper result - i try on PI ...and get a problems...

**Comment by alexzk1** [comment]
[output.txt](https://github.com/doe300/VC4CL/files/2206760/output.txt)
that was build on desktop ...well...initial warnings saw on pi... not sure if here though

**Comment by alexzk1** [comment]
I think problem rises pi is 32 bits while desktop is 64

**Comment by alexzk1** [comment]
to be honest, i pray to things - that all compiler warnings must be fixed - "trust in ir compiler"

**Comment by doe300** [comment]
> progs.build("-cl-opt-disable");

VC4C does not (yet) head this option, so they stay enabled.

I'm pretty sure, that if a calculation works fine on NVIDIA and does not on VC4CL, then the problem is on the VC4CL part.

Did you compile the binaries (VC4C and VC4CL) as 64-bit program? If so, it might be a problem, but I can't say with certainty.

**Comment by alexzk1** [comment]
no. i compiled all as packages on rpi, then git pull my prog there, the tried
all links are above

**Comment by doe300** [comment]
So I found and fixed a bug assigning locals to registers (and handling conflicts) and also found another problem, which is not yet fixed:
In several kernels, you use the index -1 for accessing memory objects. I did not really check, whether this is correct from a functionality point, but the VC4C compiler does not support negative integers. I will have to add some extra handling for this.

**Comment by alexzk1** [comment]
Yes, I add extra memory prior call t kernel, so final index must be valid in any case. If not - it is my mistake somewhere then.

**Comment by doe300** [comment]
> Also replacing by *(INPUT - 1) just in case.

That didn't do anything, since CLang converted it to the exact same instructions:)
But I now added support at least for constant negative indices, resulting in the kernels building and running in the emulator without any issues. Idk though, if the results are correct.

**Comment by alexzk1** [comment]
Okey, thanks. No hangs now on that kernel. I will test atan2 and kernels for proper results as well, but that will take some time.

**Comment by alexzk1** [comment]
Next question, this single line makes kernel 10x TIMES(!!) slower

`vstore16(select(0, Z5, isless(p2, Z5) && isless(p1, Z5)), 0, ( __global float*)(N + dstPaddedIndex));`

I would bet on vstore, bcs another lines use isless too. Any reason for that ? >: On NVIDA delay is less then time measure error.

**Comment by doe300** [comment]
This makes the kernel slower compared to what?

The otherwise same kernel with this line removed?
Then yes, it is definitively the vstore16, since memory access (esp. writing memory) is the bottleneck for VC4.

**Comment by alexzk1** [comment]
Such C++ code, prints time used inside:
```
 {
            TimeMeasure t("Kernel 2");
            kernel_non_maximum(cl::EnqueueArgs(queue, cl::NDRange(kw, kh), cl::NDRange(gpus)), bangle, bgm, bN).wait();
        }
```

So when line uncommented it is 100ms, when commented out - it is 7-8ms.

**Comment by doe300** [comment]
There is just one memory. The GPU directly accesses the CPU memory, it does not have any of its own.

For `__private` and `__local` memory, VC4C tries to optimize the memory access by either keeping the values in the registers or using the VPM (a small cache also used to write to RAM). For `__constant` memory, there is also an optimization trying to load the contents directly into registers.

`__global` memory is always located in the RAM. Also, if a memory location (the same parameter) is not written to, the loading of the data can also be sped up a bit.

On the host-side, since the memory is really shared, accessing a mapped buffer (e.g. via `clEnqueueMapBuffer`) accesses directly the physical addresses also accessed by the GPU resulting in no extra copying required.

**Comment by alexzk1** [comment]
So...there is no way to beat that ? >: Usual technique was split task into couple kernels, and keep passing buffers between them. And only final one is copied to sys memory. So, I run 3 kernels. As you see on 2nd kernel example - load all data from 1st + calculations = 8ms. Storing data to pass to 3rd = 90ms extra. I don't get why loads are much faster then storing, if it uses same unit linearly.

**Comment by alexzk1** [comment]
So what If you will check how that buffer was created, and if not accessed from host side (kinda gpu-only buffer) then you skip locks ? It is kernel's developer problem to keep all proper.

**Comment by doe300** [comment]
> So...there is no way to beat that ?

No, that is a hardware limitation. Unless of course there is some other hardware component I don't know about, or there is some trick to be done with the known components...

> I don't get why loads are much faster then storing, if it uses same unit linearly.

As I wrote in the post above, loads (if the same parameter is not written into, e.g. by `const` pointer) are optimized. They use a different hardware component, which is mainly for reading colors for texture processing, but can also read arbitrary data (and most importantly: in parallel!)

> So what If you will check how that buffer was created, [...]

How would you do that? At compilation time I don't know which memory is really accessed host-side (at least not for `__global` memory). Also, using the VPM as cache requires locking too, but for a much shorter time-span, since the data does not need to be copied into RAM under lock.

**Comment by alexzk1** [comment]
Do you take in account "restrict" keyword too ? As I understood, it tells compiler no overlaps are expected on this pointer.

..ok well, is any way to take lock once for couple vstore? For example, store 16 of float16 at once in 1 lock ?

**Comment by doe300** [comment]
> Do you take in account "restrict" keyword too ?

Currently no, I more or less assume all parameters to be restricted, which might not be right. But treating the correctly would decrease performance...

> ok well, is any way to take lock once for couple vstore?

It is done. `vstore16` writes (guarded by the hardware mutex) 16 elements first into VPM cache and then into RAM.

**Comment by alexzk1** [comment]
`It is done. vstore16 writes (guarded by the hardware mutex) 16 elements first into VPM cache and then into RAM.`

Well, I have loop there, which does vstore16 for float16 and loop itself is for 16 items. I was thinking to cache it, and store 16 of float16 (256 numbers) in 1 lock.

**Comment by doe300** [comment]
Ah okay.
There is an optimization for grouping stores into contiguous memory, but I don't think it currently works with loops, unless they are unrolled.

**Comment by alexzk1** [comment]
How big is VPM cache? What if you will do copy to ram on explicit call of cl::copy (or cache overflow)? Otherwise, if kernels are chained, you will have same address out/in and can keep using VPM then.

**Comment by alexzk1** [comment]
If I will lower amount of work items (i.e. will do deep nested loops in kernel), will it speed up vstore's somehow?

**Comment by doe300** [comment]
> How big is VPM cache?

64 rows (actually more, but only 64 can be addressed) with 16 words (32-bit) per row, but addressing works per row.

> What if you will do copy to ram on explicit call of cl::copy (or cache overflow)? 

Using the VPM as cache and only writing back when the cache is full would be a useful optimization, but also very hard to realize (esp. dynamically, e.g. with loops). The cache space needs to be split into blocks per QPU (core) and for each core the number of remaining free cache lines need to be tracked...

> Otherwise, if kernels are chained, you will have same address out/in and can keep using VPM then.

This would be very useful, but the compiler cannot know if the input comes from another kernel or the host.

> If I will lower amount of work items (i.e. will do deep nested loops in kernel), will it speed up vstore's somehow?

I doubt it, it will probably stay the same.

**Comment by alexzk1** [comment]
Damn, then it kills all usage of GPU >: 1:10 ratio of calculation/storing result is crazy.

**Comment by alexzk1** [comment]
Look here, this guy describes he gets bigger speeds:
https://github.com/mn416/QPULib

**Comment by doe300** [comment]
He introduces special functions to do asynchronous load and store.
OpenCL only provides the `async_work_group_copy` and `prefetch` functions, which both are not really what you need and also both cannot be really optimized for the VC4.

One thing though I saw in his code, or more accurately didn't see in his code, was mutex locking...

**Comment by alexzk1** [comment]
Do you use TMU too ? Is it for load only?
https://petewarden.com/2014/08/07/how-to-optimize-raspberry-pi-code-using-its-gpu/

**Comment by doe300** [comment]
Yeah, for parameters which are only read and not written to, the TMU is used.

**Comment by alexzk1** [comment]
Can you add non-opencl function, which will accept ARRAY of values or pointer (and it's size) and will do vstore in single lock? So usage can be like

#ifdef V4C
float16 tmp[16];
v4c_vstore(tmp, 16, offset, ram_pointer)
#endif

**Comment by doe300** [comment]
What exactly do you mean?
What would the function definition look like? What should be the behavior for the `vc4c_vstore`?

With the current code you may be able to do the following:

     vc4cl_mutex_lock();
    vc4cl_dma_write(ptr, data[0]);
    vc4cl_dma_write(ptr + offset, data[1]);
    vc4cl_dma_write(ptr + offset * 2, data[2]);
    vc4cl_dma_write(ptr + offset * 3, data[3]);
    // ...
    vc4cl_mutex_unlock();

This uses [internal VC4CLStdLib functions](https://github.com/doe300/VC4CLStdLib/blob/master/include/_intrinsics.h#L113), but they might change at any time without warnings!

**Comment by alexzk1** [comment]
Yes,  I was thinking if using single mutex to store many values instead 1 will speed up process. vstore takes 1 value (variable), I was thinking to use array of variables, i.e. in my example it stores 256 floats at once.

**Comment by alexzk1** [comment]
Ok, tried that:
```
 void arrayOut(float16* arr, int sz, __global float16* out, int stride)
           {
                vc4cl_mutex_lock();
                for (int i = 0; i < sz; ++i)
                {
                   __global float* ptr = (__global float*)(out + stride * i);
                   vc4cl_dma_write(ptr + 0, arr[i].s0);
                   vc4cl_dma_write(ptr + 1, arr[i].s1);
                   vc4cl_dma_write(ptr + 2, arr[i].s2);
                   vc4cl_dma_write(ptr + 3, arr[i].s3);

                   vc4cl_dma_write(ptr + 4, arr[i].s4);
                   vc4cl_dma_write(ptr + 5, arr[i].s5);
                   vc4cl_dma_write(ptr + 6, arr[i].s6);
                   vc4cl_dma_write(ptr + 7, arr[i].s7);

                   vc4cl_dma_write(ptr + 8, arr[i].s8);
                   vc4cl_dma_write(ptr + 9, arr[i].s9);
                   vc4cl_dma_write(ptr + 10, arr[i].sa);
                   vc4cl_dma_write(ptr + 11, arr[i].sb);

                   vc4cl_dma_write(ptr + 12, arr[i].sc);
                   vc4cl_dma_write(ptr + 13, arr[i].sd);
                   vc4cl_dma_write(ptr + 14, arr[i].se);
                   vc4cl_dma_write(ptr + 15, arr[i].sf);

                }
                vc4cl_mutex_unlock();
           }
```

Gives next errors:
  %call2 = tail call spir_func i32 bitcast (i32 ()* @vc4cl_dma_write to i32 (float addrspace(1)*, double)*)(float addrspace(1)* %1, double %conv) #10Failed to compiler kernels: 
[W] Sun Jul 29 18:49:21 2018: 64-bit operations are not supported by the VideoCore IV architecture, further compilation may fail!
[W] Sun Jul 29 18:49:21 2018: 64-bit operations are not supported by the VideoCore IV architecture, further compilation may fail!
[W] Sun Jul 29 18:49:21 2018: 64-bit operations are not supported by the VideoCore IV architecture, further compilation may fail!
[E] Sun Jul 29 18:49:21 2018: Compiler threw exception: Parser: Unhandled type of indirect function call!

**Comment by doe300** [comment]
I do not understand why this error is thrown, `vc4cl_dma_write` does not return anything. Also, why does it take double arguments?

Did you declare the function before use? The error looks like it was not declared, and as OpenCL is basically C, any not declared function is assumed to be of type `int ()`. You will need to declare `void vc4cl_dma_write(volatile __global float*, float)` in this case.

**Comment by alexzk1** [comment]
No, i didn't declare anything, including mutexes.
Maybe you could change texts to something more like GCC ? Like "function is implicitly declared" or so.

**Comment by doe300** [comment]
> Maybe you could change texts to something more like GCC ? Like "function is implicitly declared" or so.

For that I would need to know that the function is not declared. This is already handled by CLang by converting it to this strange bit-cast function call. From a VC4C point of view, this is just a bit-cast function call which is not supported.

> 4 times slower total, than direct vstore16

I din't notice it before, but why do you write every element of the vector extra? You could also use `vc4cl_dma_write(volatile __global float*, float16)` to write the whole vector at once.
Also, the loop with the non-constant iteration count will be a problem. If you allow clang to unroll it (e.g. by converting `sz` to a macro or constant), this should be faster than using multiple `vstore16`s.

**Comment by alexzk1** [comment]
Well, this works slightly slower (about 20%) then original vstore16 each.

```
void vc4cl_dma_write(volatile __global float*, float16);
           vc4cl_mutex_lock();
           vc4cl_mutex_unlock();
           void arrayOut16(const float16* arr,  __global float16* out, const int stride)
           {
                vc4cl_mutex_lock();
                for (int i = 0; i < 16; ++i)
                {
                   __global float* ptr = (__global float*)(out + stride * i);
                   vc4cl_dma_write(ptr + 0, arr[i]);

                }
                vc4cl_mutex_unlock();
           }
```

**Comment by doe300** [comment]
Can you send me a full example kernel of the memory access you do? So I can better see what improvements I can make that will help you.

**Comment by alexzk1** [comment]
I tried 1st kernel there - with all vstrore disabled it takes 35ms, once vstrores are enabled - takes 200+ms. Not sure why they did such chip >: just useless.
